File: flutter/lib/src/chat_template.dart
/// Chat template formatting for multi-turn conversations
///
/// Formats conversation history into model-specific prompt strings.
/// The primary format is [ChatTemplateFormat.llama3Instruct] for
/// Llama 3.x models. [ChatTemplateFormat.chatML] supports models
/// using the ChatML format. [ChatTemplateFormat.generic] provides
/// a simple fallback for unknown models.
///
/// ChatSession uses these templates to convert message history into
/// a single prompt string that the model understands as multi-turn
/// conversation.
library;

import 'chat_types.dart';

/// Supported chat template formats
///
/// Each format defines how system prompts, user messages, and assistant
/// responses are delimited in the prompt string. Using the wrong template
/// for a model will produce garbage output.
enum ChatTemplateFormat {
  /// Llama 3 Instruct format (default for Llama-3.x-Instruct models)
  ///
  /// Uses `<|begin_of_text|>`, `<|start_header_id|>`, `<|end_header_id|>`,
  /// and `<|eot_id|>` special tokens.
  llama3Instruct,

  /// ChatML format (used by many open models)
  ///
  /// Uses `<|im_start|>` and `<|im_end|>` special tokens.
  chatML,

  /// Generic fallback format using markdown-style headers
  ///
  /// Uses `### System:`, `### User:`, `### Assistant:` markers.
  /// Works as a reasonable default when the exact template is unknown.
  generic,
}

/// Formats multi-turn conversations into model-specific prompt strings
///
/// This class applies chat templates in pure Dart, avoiding the need for
/// new C API symbols. The templates match the formats expected by
/// llama.cpp's tokenizer for each model family.
///
/// Example:
/// ```dart
/// final prompt = ChatTemplate.format(
///   template: ChatTemplateFormat.llama3Instruct,
///   systemPrompt: 'You are a helpful assistant.',
///   messages: [
///     ChatMessage(role: ChatRole.user, content: 'Hello!', timestamp: DateTime.now()),
///   ],
/// );
/// ```
class ChatTemplate {
  /// Format a conversation into a prompt string using the specified template
  ///
  /// [template] determines which special tokens and delimiters to use.
  /// [systemPrompt] is optional and placed at the beginning of the prompt.
  /// [messages] is the conversation history to format.
  ///
  /// Returns a complete prompt string ready to pass to the model, ending
  /// with the assistant turn marker to prompt a response.
  static String format({
    required ChatTemplateFormat template,
    String? systemPrompt,
    required List<ChatMessage> messages,
  }) {
    switch (template) {
      case ChatTemplateFormat.llama3Instruct:
        return _formatLlama3Instruct(systemPrompt, messages);
      case ChatTemplateFormat.chatML:
        return _formatChatML(systemPrompt, messages);
      case ChatTemplateFormat.generic:
        return _formatGeneric(systemPrompt, messages);
    }
  }

  /// Format using Llama 3 Instruct template
  ///
  /// Produces:
  /// ```
  /// <|begin_of_text|><|start_header_id|>system<|end_header_id|>
  ///
  /// {system prompt}<|eot_id|><|start_header_id|>user<|end_header_id|>
  ///
  /// {user message}<|eot_id|><|start_header_id|>assistant<|end_header_id|>
  ///
  /// ```
  static String _formatLlama3Instruct(
    String? systemPrompt,
    List<ChatMessage> messages,
  ) {
    final buffer = StringBuffer();

    buffer.write('<|begin_of_text|>');

    // System prompt (optional)
    if (systemPrompt != null && systemPrompt.isNotEmpty) {
      buffer.write('<|start_header_id|>system<|end_header_id|>\n\n');
      buffer.write(systemPrompt);
      buffer.write('<|eot_id|>');
    }

    // Conversation turns
    for (final msg in messages) {
      if (msg.role == ChatRole.summary) {
        // Treat summaries as system messages with a prefix
        buffer.write('<|start_header_id|>system<|end_header_id|>\n\n');
        buffer.write('Previous conversation summary: ${msg.content}');
        buffer.write('<|eot_id|>');
      } else {
        buffer.write('<|start_header_id|>${msg.role.name}<|end_header_id|>\n\n');
        buffer.write(msg.content);
        buffer.write('<|eot_id|>');
      }
    }

    // Prompt for assistant response
    buffer.write('<|start_header_id|>assistant<|end_header_id|>\n\n');

    return buffer.toString();
  }

  /// Format using ChatML template
  ///
  /// Produces:
  /// ```
  /// <|im_start|>system
  /// {system prompt}<|im_end|>
  /// <|im_start|>user
  /// {user message}<|im_end|>
  /// <|im_start|>assistant
  /// ```
  static String _formatChatML(
    String? systemPrompt,
    List<ChatMessage> messages,
  ) {
    final buffer = StringBuffer();

    // System prompt (optional)
    if (systemPrompt != null && systemPrompt.isNotEmpty) {
      buffer.write('<|im_start|>system\n');
      buffer.write(systemPrompt);
      buffer.write('<|im_end|>\n');
    }

    // Conversation turns
    for (final msg in messages) {
      if (msg.role == ChatRole.summary) {
        // Treat summaries as system messages with a prefix
        buffer.write('<|im_start|>system\n');
        buffer.write('Previous conversation summary: ${msg.content}');
        buffer.write('<|im_end|>\n');
      } else {
        buffer.write('<|im_start|>${msg.role.name}\n');
        buffer.write(msg.content);
        buffer.write('<|im_end|>\n');
      }
    }

    // Prompt for assistant response
    buffer.write('<|im_start|>assistant\n');

    return buffer.toString();
  }

  /// Format using generic markdown-style template
  ///
  /// Produces:
  /// ```
  /// ### System:
  /// {system prompt}
  ///
  /// ### User:
  /// {user message}
  ///
  /// ### Assistant:
  /// ```
  static String _formatGeneric(
    String? systemPrompt,
    List<ChatMessage> messages,
  ) {
    final buffer = StringBuffer();

    // System prompt (optional)
    if (systemPrompt != null && systemPrompt.isNotEmpty) {
      buffer.write('### System:\n');
      buffer.write(systemPrompt);
      buffer.write('\n\n');
    }

    // Conversation turns
    for (final msg in messages) {
      if (msg.role == ChatRole.summary) {
        buffer.write('### System:\n');
        buffer.write('Previous conversation summary: ${msg.content}');
        buffer.write('\n\n');
      } else {
        // Capitalize role name for display
        final roleName = msg.role.name[0].toUpperCase() + msg.role.name.substring(1);
        buffer.write('### $roleName:\n');
        buffer.write(msg.content);
        buffer.write('\n\n');
      }
    }

    // Prompt for assistant response
    buffer.write('### Assistant:\n');

    return buffer.toString();
  }
}

File: swift/Sources/EdgeVeda/ChatTemplate.swift
import Foundation

/// Chat template formats for different model families
public enum ChatTemplate: Sendable {
    /// Llama 3 chat format
    case llama3
    
    /// ChatML format (used by OpenAI, GPT-4, etc.)
    case chatml
    
    /// Mistral/Mixtral chat format
    case mistral
    
    /// Formats a list of messages according to the template
    /// - Parameter messages: The messages to format
    /// - Returns: The formatted prompt string
    public func format(messages: [ChatMessage]) -> String {
        switch self {
        case .llama3:
            return formatLlama3(messages: messages)
        case .chatml:
            return formatChatML(messages: messages)
        case .mistral:
            return formatMistral(messages: messages)
        }
    }
    
    // MARK: - Private Formatters
    
    private func formatLlama3(messages: [ChatMessage]) -> String {
        var prompt = ""
        
        for message in messages {
            prompt += "<|start_header_id|>\(message.role.rawValue)<|end_header_id|>\n\n"
            prompt += "\(message.content)<|eot_id|>"
        }
        
        // Add assistant prompt to continue generation
        prompt += "<|start_header_id|>assistant<|end_header_id|>\n\n"
        
        return prompt
    }
    
    private func formatChatML(messages: [ChatMessage]) -> String {
        var prompt = ""
        
        for message in messages {
            prompt += "<|im_start|>\(message.role.rawValue)\n"
            prompt += "\(message.content)<|im_end|>\n"
        }
        
        // Add assistant prompt to continue generation
        prompt += "<|im_start|>assistant\n"
        
        return prompt
    }
    
    private func formatMistral(messages: [ChatMessage]) -> String {
        // Mistral Instruct v0.1 format:
        // <s>[INST] <<SYS>>\n{system}\n<</SYS>>\n\n{user} [/INST] {assistant}</s>[INST] {user2} [/INST]…
        //
        // Rules:
        // • Prompt opens with <s> (BOS token) — one per conversation, not per turn.
        // • System prompt is injected inside the first [INST] block using <<SYS>>/<</SYS>> markers.
        // • Subsequent user turns use [INST]…[/INST] without <s>.
        // • Previous implementation placed system content as bare text before [INST], which
        //   Mistral models do not recognise, and omitted the <s> BOS token entirely.
        var prompt = "<s>"
        var pendingSystem: String? = nil

        for message in messages {
            switch message.role {
            case .system:
                // Defer system content until the next user turn so it lands inside [INST].
                pendingSystem = message.content
            case .user:
                prompt += "[INST]"
                if let sys = pendingSystem {
                    prompt += " <<SYS>>\n\(sys)\n<</SYS>>\n\n"
                    pendingSystem = nil
                } else {
                    prompt += " "
                }
                prompt += "\(message.content) [/INST]"
            case .assistant:
                prompt += " \(message.content)</s>"
            }
        }

        return prompt
    }
}